lib/commands/test.ts

```diff
@@ -1,30 +1,73 @@
-	return function RunKarmaTestCommand($options: IOptions, $testExecutionService: ITestExecutionService, $projectData: IProjectData, $analyticsService: IAnalyticsService, $platformEnvironmentRequirements: IPlatformEnvironmentRequirements) {
-		$projectData.initializeProjectData();
-		$analyticsService.setShouldDispose($options.justlaunch || !$options.watch);
-		const projectFilesConfig = helpers.getProjectFilesConfig({ isReleaseBuild: $options.release });
-		this.execute = (args: string[]): Promise<void> => $testExecutionService.startKarmaServer(platform, $projectData, projectFilesConfig);
-		this.canExecute = (args: string[]): Promise<boolean> => canExecute({ $platformEnvironmentRequirements, $projectData, $options, platform });
-		this.allowedParameters = [];
-	};
-}
+abstract class TestCommandBase {
+	public allowedParameters: ICommandParameter[] = [];
+	private projectFilesConfig: IProjectFilesConfig;
+	protected abstract $projectData: IProjectData;
+	protected abstract $testExecutionService: ITestExecutionService;
+	protected abstract $analyticsService: IAnalyticsService;
+	protected abstract $options: IOptions;
+	protected abstract $platformEnvironmentRequirements: IPlatformEnvironmentRequirements;
+	protected abstract $errors: IErrors;
+
+	constructor(private platform: string) {
+	}
+
+	async execute(args: string[]): Promise<void> {
+		await this.$testExecutionService.startKarmaServer(this.platform, this.$projectData, this.projectFilesConfig);
+	}
 
-async function canExecute(input: { $platformEnvironmentRequirements: IPlatformEnvironmentRequirements, $projectData: IProjectData, $options: IOptions, platform: string }): Promise<boolean> {
-	const { $platformEnvironmentRequirements, $projectData, $options, platform } = input;
-	const output = await $platformEnvironmentRequirements.checkEnvironmentRequirements({
-		platform,
-		projectDir: $projectData.projectDir,
-		options: $options,
-		notConfiguredEnvOptions: {
-			hideSyncToPreviewAppOption: true,
-			hideCloudBuildOption: true
+	async canExecute(args: string[]): Promise<boolean | ICanExecuteCommandOutput> {
+		this.$projectData.initializeProjectData();
+		this.$analyticsService.setShouldDispose(this.$options.justlaunch || !this.$options.watch);
+		this.projectFilesConfig = helpers.getProjectFilesConfig({ isReleaseBuild: this.$options.release });
+
+		const output = await this.$platformEnvironmentRequirements.checkEnvironmentRequirements({
+			platform: this.platform,
+			projectDir: this.$projectData.projectDir,
+			options: this.$options,
+			notConfiguredEnvOptions: {
+				hideSyncToPreviewAppOption: true,
+				hideCloudBuildOption: true
+			}
+		});
+
+		const canStartKarmaServer = await this.$testExecutionService.canStartKarmaServer(this.$projectData);
+		if (!canStartKarmaServer) {
+			this.$errors.fail({
+				formatStr: "Error: In order to run unit tests, your project must already be configured by running $ tns test init.",
+				suppressCommandHelp: true,
```

Comment:
Maybe we can create better user experience by prompt the user to automatically execute `tns test init` command.
This is not a merge stopper and can be done in another PR if we decide it is appropriate.

Reply:
https://github.com/NativeScript/nativescript-cli/issues/4387